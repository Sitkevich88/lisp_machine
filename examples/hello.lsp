(setq a "hello")
(print a)