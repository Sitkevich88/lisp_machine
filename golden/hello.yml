source: |-
  (print "Hello world!")
input: |

code: |-
  [
      {
          "opcode": "loadc",
          "arg": 72,
          "term": 0
      },
      {
          "opcode": "store",
          "arg": 0,
          "term": 1
      },
      {
          "opcode": "loadc",
          "arg": 101,
          "term": 2
      },
      {
          "opcode": "store",
          "arg": 1,
          "term": 3
      },
      {
          "opcode": "loadc",
          "arg": 108,
          "term": 4
      },
      {
          "opcode": "store",
          "arg": 2,
          "term": 5
      },
      {
          "opcode": "loadc",
          "arg": 108,
          "term": 6
      },
      {
          "opcode": "store",
          "arg": 3,
          "term": 7
      },
      {
          "opcode": "loadc",
          "arg": 111,
          "term": 8
      },
      {
          "opcode": "store",
          "arg": 4,
          "term": 9
      },
      {
          "opcode": "loadc",
          "arg": 32,
          "term": 10
      },
      {
          "opcode": "store",
          "arg": 5,
          "term": 11
      },
      {
          "opcode": "loadc",
          "arg": 119,
          "term": 12
      },
      {
          "opcode": "store",
          "arg": 6,
          "term": 13
      },
      {
          "opcode": "loadc",
          "arg": 111,
          "term": 14
      },
      {
          "opcode": "store",
          "arg": 7,
          "term": 15
      },
      {
          "opcode": "loadc",
          "arg": 114,
          "term": 16
      },
      {
          "opcode": "store",
          "arg": 8,
          "term": 17
      },
      {
          "opcode": "loadc",
          "arg": 108,
          "term": 18
      },
      {
          "opcode": "store",
          "arg": 9,
          "term": 19
      },
      {
          "opcode": "loadc",
          "arg": 100,
          "term": 20
      },
      {
          "opcode": "store",
          "arg": 10,
          "term": 21
      },
      {
          "opcode": "loadc",
          "arg": 33,
          "term": 22
      },
      {
          "opcode": "store",
          "arg": 11,
          "term": 23
      },
      {
          "opcode": "loadc",
          "arg": 0,
          "term": 24
      },
      {
          "opcode": "store",
          "arg": 12,
          "term": 25
      },
      {
          "opcode": "loadc",
          "arg": 3,
          "term": 26
      },
      {
          "opcode": "store",
          "arg": 13,
          "term": 27
      },
      {
          "opcode": "loadc",
          "arg": 0,
          "term": 28
      },
      {
          "opcode": "store",
          "arg": 14,
          "term": 29
      },
      {
          "opcode": "load",
          "arg": 14,
          "term": 30
      },
      {
          "opcode": "store",
          "arg": 15,
          "term": 31
      },
      {
          "opcode": "load",
          "arg": 13,
          "term": 32
      },
      {
          "opcode": "je",
          "arg": 47,
          "term": 33
      },
      {
          "opcode": "addc",
          "arg": -1,
          "term": 34
      },
      {
          "opcode": "je",
          "arg": 57,
          "term": 35
      },
      {
          "opcode": "addc",
          "arg": -1,
          "term": 36
      },
      {
          "opcode": "je",
          "arg": 45,
          "term": 37
      },
      {
          "opcode": "load_indir",
          "arg": 15,
          "term": 38
      },
      {
          "opcode": "je",
          "arg": 59,
          "term": 39
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 40
      },
      {
          "opcode": "load",
          "arg": 15,
          "term": 41
      },
      {
          "opcode": "addc",
          "arg": 1,
          "term": 42
      },
      {
          "opcode": "store",
          "arg": 15,
          "term": 43
      },
      {
          "opcode": "jmp",
          "arg": 38,
          "term": 44
      },
      {
          "opcode": "load",
          "arg": 15,
          "term": 45
      },
      {
          "opcode": "jne",
          "arg": 54,
          "term": 46
      },
      {
          "opcode": "loadc",
          "arg": 78,
          "term": 47
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 48
      },
      {
          "opcode": "loadc",
          "arg": 73,
          "term": 49
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 50
      },
      {
          "opcode": "loadc",
          "arg": 76,
          "term": 51
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 52
      },
      {
          "opcode": "jmp",
          "arg": 59,
          "term": 53
      },
      {
          "opcode": "loadc",
          "arg": 116,
          "term": 54
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 55
      },
      {
          "opcode": "jmp",
          "arg": 59,
          "term": 56
      },
      {
          "opcode": "load",
          "arg": 15,
          "term": 57
      },
      {
          "opcode": "print_int",
          "arg": "",
          "term": 58
      },
      {
          "opcode": "loadc",
          "arg": 10,
          "term": 59
      },
      {
          "opcode": "print",
          "arg": "",
          "term": 60
      },
      {
          "opcode": "halt",
          "arg": "",
          "term": 61
      }
  ]

output: |
  source LoC: 3 , code instr: 62
  ============================================================
  Hello world!

  instr_counter:  126 ticks: 152

log: |+
  DEBUG machine:simulation {TICK: 0, PC: 0, ADDR: 0, OUT: 0, ACC: 0}, {OPCODE: loadc, ARG: 72, POS: 0}
  DEBUG machine:simulation {TICK: 1, PC: 1, ADDR: 0, OUT: 0, ACC: 72}, {OPCODE: store, ARG: 0, POS: 1}
  DEBUG machine:simulation {TICK: 2, PC: 2, ADDR: 0, OUT: 72, ACC: 72}, {OPCODE: loadc, ARG: 101, POS: 2}
  DEBUG machine:simulation {TICK: 3, PC: 3, ADDR: 0, OUT: 72, ACC: 101}, {OPCODE: store, ARG: 1, POS: 3}
  DEBUG machine:simulation {TICK: 4, PC: 4, ADDR: 1, OUT: 101, ACC: 101}, {OPCODE: loadc, ARG: 108, POS: 4}
  DEBUG machine:simulation {TICK: 5, PC: 5, ADDR: 1, OUT: 101, ACC: 108}, {OPCODE: store, ARG: 2, POS: 5}
  DEBUG machine:simulation {TICK: 6, PC: 6, ADDR: 2, OUT: 108, ACC: 108}, {OPCODE: loadc, ARG: 108, POS: 6}
  DEBUG machine:simulation {TICK: 7, PC: 7, ADDR: 2, OUT: 108, ACC: 108}, {OPCODE: store, ARG: 3, POS: 7}
  DEBUG machine:simulation {TICK: 8, PC: 8, ADDR: 3, OUT: 108, ACC: 108}, {OPCODE: loadc, ARG: 111, POS: 8}
  DEBUG machine:simulation {TICK: 9, PC: 9, ADDR: 3, OUT: 108, ACC: 111}, {OPCODE: store, ARG: 4, POS: 9}
  DEBUG machine:simulation {TICK: 10, PC: 10, ADDR: 4, OUT: 111, ACC: 111}, {OPCODE: loadc, ARG: 32, POS: 10}
  DEBUG machine:simulation {TICK: 11, PC: 11, ADDR: 4, OUT: 111, ACC: 32}, {OPCODE: store, ARG: 5, POS: 11}
  DEBUG machine:simulation {TICK: 12, PC: 12, ADDR: 5, OUT: 32, ACC: 32}, {OPCODE: loadc, ARG: 119, POS: 12}
  DEBUG machine:simulation {TICK: 13, PC: 13, ADDR: 5, OUT: 32, ACC: 119}, {OPCODE: store, ARG: 6, POS: 13}
  DEBUG machine:simulation {TICK: 14, PC: 14, ADDR: 6, OUT: 119, ACC: 119}, {OPCODE: loadc, ARG: 111, POS: 14}
  DEBUG machine:simulation {TICK: 15, PC: 15, ADDR: 6, OUT: 119, ACC: 111}, {OPCODE: store, ARG: 7, POS: 15}
  DEBUG machine:simulation {TICK: 16, PC: 16, ADDR: 7, OUT: 111, ACC: 111}, {OPCODE: loadc, ARG: 114, POS: 16}
  DEBUG machine:simulation {TICK: 17, PC: 17, ADDR: 7, OUT: 111, ACC: 114}, {OPCODE: store, ARG: 8, POS: 17}
  DEBUG machine:simulation {TICK: 18, PC: 18, ADDR: 8, OUT: 114, ACC: 114}, {OPCODE: loadc, ARG: 108, POS: 18}
  DEBUG machine:simulation {TICK: 19, PC: 19, ADDR: 8, OUT: 114, ACC: 108}, {OPCODE: store, ARG: 9, POS: 19}
  DEBUG machine:simulation {TICK: 20, PC: 20, ADDR: 9, OUT: 108, ACC: 108}, {OPCODE: loadc, ARG: 100, POS: 20}
  DEBUG machine:simulation {TICK: 21, PC: 21, ADDR: 9, OUT: 108, ACC: 100}, {OPCODE: store, ARG: 10, POS: 21}
  DEBUG machine:simulation {TICK: 22, PC: 22, ADDR: 10, OUT: 100, ACC: 100}, {OPCODE: loadc, ARG: 33, POS: 22}
  DEBUG machine:simulation {TICK: 23, PC: 23, ADDR: 10, OUT: 100, ACC: 33}, {OPCODE: store, ARG: 11, POS: 23}
  DEBUG machine:simulation {TICK: 24, PC: 24, ADDR: 11, OUT: 33, ACC: 33}, {OPCODE: loadc, ARG: 0, POS: 24}
  DEBUG machine:simulation {TICK: 25, PC: 25, ADDR: 11, OUT: 33, ACC: 0}, {OPCODE: store, ARG: 12, POS: 25}
  DEBUG machine:simulation {TICK: 26, PC: 26, ADDR: 12, OUT: 0, ACC: 0}, {OPCODE: loadc, ARG: 3, POS: 26}
  DEBUG machine:simulation {TICK: 27, PC: 27, ADDR: 12, OUT: 0, ACC: 3}, {OPCODE: store, ARG: 13, POS: 27}
  DEBUG machine:simulation {TICK: 28, PC: 28, ADDR: 13, OUT: 3, ACC: 3}, {OPCODE: loadc, ARG: 0, POS: 28}
  DEBUG machine:simulation {TICK: 29, PC: 29, ADDR: 13, OUT: 3, ACC: 0}, {OPCODE: store, ARG: 14, POS: 29}
  DEBUG machine:simulation {TICK: 30, PC: 30, ADDR: 14, OUT: 0, ACC: 0}, {OPCODE: load, ARG: 14, POS: 30}
  DEBUG machine:simulation {TICK: 31, PC: 31, ADDR: 14, OUT: 0, ACC: 0}, {OPCODE: store, ARG: 15, POS: 31}
  DEBUG machine:simulation {TICK: 32, PC: 32, ADDR: 15, OUT: 0, ACC: 0}, {OPCODE: load, ARG: 13, POS: 32}
  DEBUG machine:simulation {TICK: 33, PC: 33, ADDR: 13, OUT: 3, ACC: 3}, {OPCODE: je, ARG: 47, POS: 33}
  DEBUG machine:simulation {TICK: 34, PC: 34, ADDR: 13, OUT: 3, ACC: 3}, {OPCODE: addc, ARG: -1, POS: 34}
  DEBUG machine:simulation {TICK: 35, PC: 35, ADDR: 13, OUT: 3, ACC: 2}, {OPCODE: je, ARG: 57, POS: 35}
  DEBUG machine:simulation {TICK: 36, PC: 36, ADDR: 13, OUT: 3, ACC: 2}, {OPCODE: addc, ARG: -1, POS: 36}
  DEBUG machine:simulation {TICK: 37, PC: 37, ADDR: 13, OUT: 3, ACC: 1}, {OPCODE: je, ARG: 45, POS: 37}
  DEBUG machine:simulation {TICK: 38, PC: 38, ADDR: 13, OUT: 3, ACC: 1}, {OPCODE: load_indir, ARG: 15, POS: 38}
  DEBUG machine:simulation {TICK: 41, PC: 39, ADDR: 0, OUT: 72, ACC: 72}, {OPCODE: je, ARG: 59, POS: 39}
  DEBUG machine:simulation {TICK: 42, PC: 40, ADDR: 0, OUT: 72, ACC: 72}, {OPCODE: print, ARG: , POS: 40}
  DEBUG machine:output output: '' << 'H'
  DEBUG machine:simulation {TICK: 43, PC: 41, ADDR: 0, OUT: 72, ACC: 72}, {OPCODE: load, ARG: 15, POS: 41}
  DEBUG machine:simulation {TICK: 44, PC: 42, ADDR: 15, OUT: 0, ACC: 0}, {OPCODE: addc, ARG: 1, POS: 42}
  DEBUG machine:simulation {TICK: 45, PC: 43, ADDR: 15, OUT: 0, ACC: 1}, {OPCODE: store, ARG: 15, POS: 43}
  DEBUG machine:simulation {TICK: 46, PC: 44, ADDR: 15, OUT: 1, ACC: 1}, {OPCODE: jmp, ARG: 38, POS: 44}
  DEBUG machine:simulation {TICK: 47, PC: 38, ADDR: 15, OUT: 1, ACC: 1}, {OPCODE: load_indir, ARG: 15, POS: 38}
  DEBUG machine:simulation {TICK: 50, PC: 39, ADDR: 1, OUT: 101, ACC: 101}, {OPCODE: je, ARG: 59, POS: 39}
  DEBUG machine:simulation {TICK: 51, PC: 40, ADDR: 1, OUT: 101, ACC: 101}, {OPCODE: print, ARG: , POS: 40}
  DEBUG machine:output output: 'H' << 'e'
  DEBUG machine:simulation {TICK: 52, PC: 41, ADDR: 1, OUT: 101, ACC: 101}, {OPCODE: load, ARG: 15, POS: 41}
  DEBUG machine:simulation {TICK: 53, PC: 42, ADDR: 15, OUT: 1, ACC: 1}, {OPCODE: addc, ARG: 1, POS: 42}
  DEBUG machine:simulation ...
  DEBUG machine:output output: 'He' << 'l'
  DEBUG machine:output output: 'Hel' << 'l'
  DEBUG machine:output output: 'Hell' << 'o'
  DEBUG machine:output output: 'Hello' << ' '
  DEBUG machine:output output: 'Hello ' << 'w'
  DEBUG machine:output output: 'Hello w' << 'o'
  DEBUG machine:output output: 'Hello wo' << 'r'
  DEBUG machine:output output: 'Hello wor' << 'l'
  DEBUG machine:output output: 'Hello worl' << 'd'
  DEBUG machine:output output: 'Hello world' << '!'
  DEBUG machine:output output: 'Hello world!' << '\n'
  INFO machine:simulation output_buffer: Hello world!

...
